tensorflow/contrib/data/kernels/unordered_merge_dataset_op.cc

```diff
@@ -0,0 +1,218 @@
+#include "tensorflow/core/framework/tensor.h"
+
+namespace tensorflow {
+
+namespace {
+
+// See documentation in ../ops/dataset_ops.cc for a high-level
+// description of the following op.
+
+// This is based on zip_dataset_op.cc
+class UnorderedMergeDatasetOp : public DatasetOpKernel {
+ public:
+  explicit UnorderedMergeDatasetOp(OpKernelConstruction* ctx)
+      : DatasetOpKernel(ctx) {}
+
+  void MakeDataset(OpKernelContext* ctx, DatasetBase** output) override {
+    std::vector<DatasetBase*> inputs;
+    DatasetBase* input0;
+    OP_REQUIRES_OK(ctx, GetDatasetFromVariantTensor(ctx->input(0), &input0));
+    inputs.push_back(input0);
+    for (size_t i = 1; i < ctx->num_inputs(); ++i) {
+      DatasetBase* input;
+      OP_REQUIRES_OK(ctx, GetDatasetFromVariantTensor(ctx->input(i), &input));
+      // from VerifyTypesMatch()
+      OP_REQUIRES(
+          ctx, input0->output_dtypes().size() == input->output_dtypes().size(),
+          errors::InvalidArgument(
+              "Number of components does not match: expected ",
+              input0->output_dtypes().size(), " types but got ",
+              input->output_dtypes().size(), "."));
+      for (size_t i = 0; i < input0->output_dtypes().size(); ++i) {
+        OP_REQUIRES(ctx,
+                    input0->output_dtypes()[i] == input->output_dtypes()[i],
+                    errors::InvalidArgument(
+                        "Data type mismatch at component ", i, ": expected ",
+                        DataTypeString(input0->output_dtypes()[i]), " but got ",
+                        DataTypeString(input->output_dtypes()[i]), "."));
+      }
+      // from VerifyShapesCompatible()
+      OP_REQUIRES(
+          ctx, input0->output_shapes().size() == input->output_shapes().size(),
+          errors::InvalidArgument(
+              "Number of components does not match: expected ",
+              input0->output_shapes().size(), " types but got ",
+              input->output_shapes().size(), "."));
+      for (size_t i = 0; i < input0->output_shapes().size(); ++i) {
+        OP_REQUIRES(ctx, input0->output_shapes()[i].IsCompatibleWith(
+                             input->output_shapes()[i]),
+                    errors::InvalidArgument(
+                        "Incompatible shapes at component ", i, ": expected ",
+                        input0->output_shapes()[i].DebugString(), " but got ",
+                        input->output_shapes()[i].DebugString(), "."));
+      }
+      inputs.push_back(input);
+    }
+    *output = new Dataset(ctx, inputs);
+  }
+
+ private:
+  class Dataset : public GraphDatasetBase {
+   public:
+    explicit Dataset(OpKernelContext* ctx,
+                     const std::vector<DatasetBase*>& inputs)
+        : GraphDatasetBase(ctx), inputs_(inputs) {
+      const auto& input = inputs_[0];
+      for (DataType dt : input->output_dtypes()) {
+        output_dtypes_.push_back(dt);
+      }
+      output_shapes_.insert(output_shapes_.end(),
```

Comment:
The shape for each component here should be the "most specific compatible shape" of all the inputs. See the logic in `concatenate_dataset_op.cc` for an example:

https://github.com/tensorflow/tensorflow/blob/4984a60e7147edef532ca1b15050471e81e45841/tensorflow/core/kernels/data/concatenate_dataset_op.cc#L167